ambry-tools/src/main/java/com.github.ambry/tools/admin/ConcurrencyTestTool.java

```diff
@@ -0,0 +1,975 @@
+            }
+            while (buf.hasRemaining()) {
+              if (putContentBuf.get() != buf.get()) {
+                exceptionToReturn = new IllegalStateException("Get and Put blob content mis-match for " + blobId);
+              }
+              readBytes++;
+            }
+            getChannel.resolveOldestChunk(null);
+          } while (readBytes < bytesToRead);
+          if (readBytes != readIntoFuture.get()) {
+            exceptionToReturn = new IllegalStateException(
+                "The returned length in the future " + readIntoFuture.get() + " didn't match the "
+                    + "length of data written " + readBytes + " for " + blobId);
+          }
+          if (getChannel.getNextChunk(0) != null) {
+            exceptionToReturn = new IllegalStateException(
+                "More data found in the channel after reading all the required bytes for " + blobId);
+          }
+          readIntoFuture.get();
+        } catch (InterruptedException e) {
+          exceptionToReturn = e;
+        } catch (ExecutionException e) {
+          exceptionToReturn = e;
+        } finally {
+          futureResult.done(null, exceptionToReturn);
+          if (callback != null) {
+            callback.onCompletion(null, exceptionToReturn);
+          }
+        }
+      }
+    }
+
+    /**
+     * {@inheritDoc}
+     * Request to delete the blob from the server directly pertaining to the {@code blobId} passed in and verify that
+     * the subsequent get fails with {@link RouterErrorCode#BlobDeleted}
+     * @param blobId the blobId the of that blob that needs to be fetched
+     * @param expectedErrorCode the expected error code for the Delete call
+     * @param futureResult the {@link FutureResult} where the result/exception will be set on completion
+     * @param callback the {@link Callback} that will be called on completion
+     */
+    @Override
+    public void deleteBlobAndValidate(final String blobId, final RouterErrorCode expectedErrorCode,
+        final FutureResult futureResult, final Callback callback) {
+
+      try {
+        final Long startTimeGetBlob = SystemTime.getInstance().nanoseconds();
+        router.deleteBlob(blobId, new Callback<Void>() {
+          @Override
+          public void onCompletion(Void result, Exception exception) {
+            long latencyPerBlob = SystemTime.getInstance().nanoseconds() - startTimeGetBlob;
+            logger.trace(" Time taken to delete blob " + blobId + " in ms " + latencyPerBlob / SystemTime.NsPerMs);
+            final AtomicReference<Exception> exceptionToReturn = new AtomicReference<Exception>();
+            if (exception == null) {
+              logger.trace("Deletion of " + blobId + " succeeded. Issuing Get to confirm the deletion ");
+              FutureResult<GetBlobResult> getBlobResultFutureResult = new FutureResult<GetBlobResult>();
+              getBlobAndValidate(blobId, null, RouterErrorCode.BlobDeleted, getBlobResultFutureResult,
```

Comment:
Since this was designed right now to test a single server, I would like to wait until we use the tool to test the frontends and add this fix if need be. But good point that you brought this up.